File: Android/app/src/main/java/com/hangten/android/report/HangboardIssueReport.kt
package com.hangten.android.report

import android.net.Uri
import com.hangten.android.content.Board
import com.hangten.android.content.BoardPresentation
import com.hangten.training.BuildConfig

class HangboardIssueReportContext private constructor(
    val boardID: String,
    val boardName: String,
    val manufacturer: String,
    val presentationID: String,
    val presentationName: String,
    val platform: String,
    val appVersion: String,
    val build: String,
) {
    constructor(
        board: Board,
        presentation: BoardPresentation,
        appVersion: String = BuildConfig.VERSION_NAME,
        build: String = BuildConfig.VERSION_CODE.toString(),
    ) : this(
        boardID = board.id,
        boardName = board.name,
        manufacturer = board.manufacturer,
        presentationID = presentation.id,
        presentationName = presentation.name,
        platform = "Android",
        appVersion = appVersion,
        build = build,
    )
}

object HangboardIssueReportUrl {
    fun make(formUrl: String, context: HangboardIssueReportContext): Uri? {
        val formUri = validatedFormUri(formUrl) ?: return null

        return formUri.buildUpon()
            .clearQuery()
            .appendQueryParameter("board_id", context.boardID)
            .appendQueryParameter("board_name", context.boardName)
            .appendQueryParameter("manufacturer", context.manufacturer)
            .appendQueryParameter("presentation_id", context.presentationID)
            .appendQueryParameter("presentation_name", context.presentationName)
            .appendQueryParameter("platform", context.platform)
            .appendQueryParameter("app_version", context.appVersion)
            .appendQueryParameter("build", context.build)
            .build()
    }

    private fun validatedFormUri(rawValue: String): Uri? {
        if (rawValue.isBlank()) return null

        val uri = runCatching { Uri.parse(rawValue) }.getOrNull() ?: return null
        val host = uri.host?.lowercase() ?: return null
        return uri.takeIf {
            uri.scheme.equals("https", ignoreCase = true) &&
                uri.userInfo == null &&
                (host == "tally.so" || host.endsWith(".tally.so"))
        }
    }
}

File: HangTen/Models/HangboardIssueReport.swift
import Foundation

struct HangboardIssueReportContext: Equatable {
    let boardID: String
    let boardName: String
    let manufacturer: String
    let presentationID: String
    let presentationName: String
    let platform: String
    let appVersion: String
    let build: String

    init(
        board: TrainingBoard,
        presentation: BoardPresentation,
        appVersion: String,
        build: String
    ) {
        boardID = board.id
        boardName = board.name
        manufacturer = board.manufacturer
        presentationID = presentation.id
        presentationName = presentation.name
        platform = "iOS"
        self.appVersion = appVersion
        self.build = build
    }

    init(
        board: TrainingBoard,
        presentation: BoardPresentation,
        bundle: Bundle = .main
    ) {
        self.init(
            board: board,
            presentation: presentation,
            appVersion: bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
                ?? "unknown",
            build: bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String
                ?? "unknown"
        )
    }
}

enum HangboardIssueReportConfiguration {
    static func formURL(bundle: Bundle = .main) -> URL? {
        formURL(rawValue: bundle.object(forInfoDictionaryKey: "HANGBOARD_REPORT_FORM_URL") as? String)
    }

    static func formURL(rawValue: String?) -> URL? {
        guard let rawValue,
              !rawValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let components = URLComponents(string: rawValue),
              components.scheme?.lowercased() == "https",
              components.user == nil,
              components.password == nil,
              let host = components.host?.lowercased(),
              host == "tally.so" || host.hasSuffix(".tally.so")
        else {
            return nil
        }

        return components.url
    }
}

enum HangboardIssueReportURL {
    static func make(formURL: URL, context: HangboardIssueReportContext) -> URL? {
        guard HangboardIssueReportConfiguration.formURL(rawValue: formURL.absoluteString) != nil,
              var components = URLComponents(url: formURL, resolvingAgainstBaseURL: false)
        else {
            return nil
        }

        components.queryItems = [
            URLQueryItem(name: "board_id", value: context.boardID),
            URLQueryItem(name: "board_name", value: context.boardName),
            URLQueryItem(name: "manufacturer", value: context.manufacturer),
            URLQueryItem(name: "presentation_id", value: context.presentationID),
            URLQueryItem(name: "presentation_name", value: context.presentationName),
            URLQueryItem(name: "platform", value: context.platform),
            URLQueryItem(name: "app_version", value: context.appVersion),
            URLQueryItem(name: "build", value: context.build)
        ]
        return components.url
    }
}
